### ISSUE TITLE
python3 macosx Malloc error when entering any shell command?

### ISSUE THREAD

**Issue by guxuanzhen16** [NYUCCL/psiTurk#371]
After opening psiturk, server on yields the following error:

```
[psiTurk server:off mode:sdbx #HITs:0]$ server on
Python(80217,0x1108bc5c0) malloc: *** error for object 0x109f05040: pointer being freed was not allocated
Python(80217,0x1108bc5c0) malloc: *** set a breakpoint in malloc_error_break to debug
Abort trap: 6
```

I've tried using virtualenv to run psiturk in both Python 2.7.16 and 3.7.4 but the error is the same. Thanks in advance.

**Comment by Xinzhu-Fang** [comment]
![image](https://user-images.githubusercontent.com/29473163/62571872-6f964100-b860-11e9-8f01-2d7f2a3a0766.png)
having the same issue on my brand new mac (unboxed yesterday). psiturk 2.3.3; python 3.7; `pip install`ed through anaconda 3. 
Help please

**Comment by Xinzhu-Fang** [comment]
this error disappeared for me when I switched to a conda env in python2.7

**Comment by guxuanzhen16** [comment]
In Conda psiturk would still try to run on python3 for some reason, so I had to manual edit each file to run python2.7, which solved the problem for me. It seems psiturk is still not working very steadily on python3.

**Comment by deargle** [comment]
It works fine on python3 outside of conda -- e.g., in a virtualenv. Passes
tests on travis. I don't support conda.

**Comment by tchamberlain** [comment]
I was also getting this error using python3 in a virtualenv, i switched to using python 2.7 and the error went away

**Comment by deargle** [comment]
Is the common thread that everyone here is using conda?

**Comment by guxuanzhen16** [comment]
Before I tried conda I just directly used python3, which did not work either. It seems that it worked for everyone after switching back to 2.7

**Comment by deargle** [comment]
Right, but I want to solve the python3 part, since python2 is EOL end of
this year.

**Comment by Rick-C-137** [comment]
I'm also having the same issue on python 3, but have had success with python 2.7. Do you have any insight into why conda creates a problem here?

**Comment by deargle** [comment]
I can't replicate the error, so I don't have insight. Can you (or anyone experiencing the error) give me as many gory details of your setup as possible? I don't know exactly what to ask for, but maybe what python version _and_ distribution you have installed _and_ how you installed psiturk. Again, psiturk works fine with python 2.7, python 3.6 and python 3.7 on travis, and travis uses a clean environment. See e.g. https://travis-ci.org/NYUCCL/psiTurk/builds/589276088 

@Xinzhu-Fang up above mentioned that they're using anaconda 3 on a mac. I'd like to maybe see:
* exact steps for setting up any virtual environment
* steps for installing psiturk

I'm guessing wildly that it's something to do with anaconda-python being used,.

**Comment by Rick-C-137** [comment]
Currently checking with a clean (non-conda) version. I'll let you know.

**Comment by deargle** [comment]
fyi, travis tests against the following python versions:
* 2.7.15
* 3.6.7
* 3.7.1

**Comment by Rick-C-137** [comment]
I should point out that Xinzhu-Fang and I are both working on OSX. The OP might also be, thought they didn't mention. I assume you're using linux, since the travis docs said it is not supported for mac.

**Comment by deargle** [comment]
Yes, I do all development on a linux (mint) virtual machine. I download my python versions from pyenv.

**Comment by deargle** [comment]
I just installed anaconda3 into my linux distro and tested psiturk v2.3.5 using anaconda's python3.7.3, which ran without errors.

**Comment by deargle** [comment]
@Rick-C-137 where were you reading about travis not supporting mac? https://docs.travis-ci.com/user/reference/osx/

**Comment by Rick-C-137** [comment]
I have no trouble on my ubuntu 18 server either with conda. But my laptop is mac, so I want to be able to debug on it. I haven't had success yet in testing if the issue on OSX derives from anaconda, because I'm running into dependency issues with system python installs, which I haven't been able to resolve yet (something to do with not being able to find/use gnureadline).

The red box here: https://docs.travis-ci.com/user/languages/python/
I don't really know anything about travis, so not sure if I'm understanding correctly...

**Comment by deargle** [comment]
I successfully set up travis to test against osx 10.14.4. No errors from the unit tests. Which leads me to suspect gnureadline. The tests give a compiler warning about it here: https://travis-ci.org/NYUCCL/psiTurk/jobs/590147471#L300 And, the unit tests aren't testing readline input.

**Comment by deargle** [comment]
I noticed that gnureadline was pinned to 6.3.3 in setup.py -- v8.0.0 is now available. But, perhaps it should be up to the user if they want to use gnureadline over whatever default readline version python distro includes. See the big "First... STOP" info at the top of the gnureadline package page. https://pypi.org/project/gnureadline/

I'm betting (hoping) that gnureadline==8.0.0 will fix macosx user woes. But can one (all?) of you try installing psiturk from the head of the `python2` branch, _using python 3_, and tell me if the shell works now? e.g.,

`mkvirtualenv -p <path to some version of python3>`
`git clone git@github.com:NYUCCL/psiTurk.git`
`cd psiTurk`
`git checkout python2`
`pip install -e .`

Then, navigate to your psiturk project dir and try launching the shell, and then try running any command. I noticed up above that `server on` and `version` both threw the error, so probably any command will throw the error.

Also, @fredcallaway you mentioned something about gnureadline incompatibilities, maybe this will fix things for you, too.

**Comment by fredcallaway** [comment]
Following these steps I get an error: `ModuleNotFoundError: No module named 'psycopg2'`

After pip installing, I can start up the shell and post a sandbox hit. No issues with readline!

**Comment by deargle** [comment]
Great! Would be good if others on the thread can chip in too.

Psycopg2 is required if using a psql db connection, but psiturk lets users
use whatever db they want (mysql, psql). Maybe i'll catch and print out an
error if psql is in url but if psycopg2 not found.

**Comment by Rick-C-137** [comment]
I followed your steps above, and got gunicorn errors with the python2 branch. Then I switched back to master and everything worked perfectly. I'm working on homebrew, which I finally got working properly again, so it seems that master works on homebrew python3 but not anaconda. Thanks for your help. Any guesses as to what kind of thing would be different in anaconda vs. homebrew?

**Comment by deargle** [comment]
Can you show the gunicorn errors when using python 3 (both anaconda and
homebrew) on the python2 branch? Should be in server.log. the python2
branch supports both python 3 and python 2, while master is python3 only.
So master branch has some compatibility code stripped out, and uses a more
recent cmd2 library which is python3 only, which might be fixing whatever
was breaking for old gnureadline versions in this thread. Python2 branch is
basically stuck on using legacy cmd2 code, which is what is used for the
shell.

Be aware that master branch has default enabled a webapi that can be used
to post hits, approve and bonus workers, etc. Won't accept default username
password from config.txt though.

I didnt do anything different with the master branch yesterday regarding
gnureadline. But if you look at that pip gnureadline page, it says that
different python distributions (e.g. homebrew, anaconda) come with
different static linked versions of readline, which is compiled c, which is
what was throwing the malloc errors I think.

**Comment by Rick-C-137** [comment]
I'm see this error in the log:
ImportError: cannot import name 'InvalidUsageError' from 'psiturk.experiment_errors'

Not sure what it's about...I can give you the whole printout if you'd like.

**Comment by deargle** [comment]
Ah, yes. Rerun psiturk-setup-example if you are using psiturk from master
branch. Some class names changed in custom.py from the python2 branch. Not
related to the gnureadline malloc error.

But if you did not get a malloc error when trying to run a command such as
`version` from a psiturk python2-branch shell using python 3, i'll close
this as resolved.

**Comment by Rick-C-137** [comment]
The `version` command works, but `server on` fails.

**Comment by deargle** [comment]
What! Fails with the malloc error specifically, even after using a fresh
project dir from a psiturk-setup-example command?

**Comment by Rick-C-137** [comment]
No, sorry for being unclear. It was that gunicorn thing (maybe because I didn't remake the example experiment).

**Comment by deargle** [comment]
Yes, absolutely because you didn't remake the example experiment / because your custom.py generated under psiturk v2 is incompatible with psiturk v3. Thanks for reporting back about malloc being gone! I'm closing this.

TLDR to this whole thread -- psiturk was installing a very old gnureadline that doesn't play nice with python3 on macosx. Head of python2 branch no longer forces install of gnureadline for macosx. If you need it, install it from pip yourself. But most mac osx python distros don't need it.
